torchao/dtypes/nf4tensor.py

```diff
@@ -616,9 +972,64 @@ def function_to_dtype(*args, **kwargs):
     if isinstance(args[0], NF4Tensor) and isinstance(args[1], torch.dtype):
         # Tensor.to(dtype, non_blocking, copy, memory_format)
         return args[0].get_original_weight().to(*args[1:], **kwargs)
+    elif isinstance(args[0], NF4Tensor) and (
+        isinstance(args[1], torch.device) or (
+            isinstance(args[1], str) and (
+                args[1] == "cpu" or args[1].startswith("cuda")
+            )
+        )
+    ) and len(args) == 2:
+        # Tensor.to(device, non_blocking)
+        device = args[1]
+        quantized_scalers = args[0].quantized_scalers.to(*(args[1:]), **kwargs)
+        quantization_factor = args[0].quantization_factor.to(*(args[1:]), **kwargs)
+        quantized_data = args[0].quantized_data.to(*(args[1:]), **kwargs)
+        return NF4Tensor(
+            SubclassTensorArgs(
+                args[0].size(),
+                args[0].stride(),
+                args[0].storage_offset(),
+                args[0].dtype,
+                device,
+                args[0].requires_grad,
+            ),
+            args[0].block_size,
+            args[0].n_blocks,
+            args[0].scaler_block_size,
+            quantized_scalers,
+            quantization_factor,
+            args[0].scaler_mean,
+            quantized_data,
+            args[0].nf4,
+        )
     else:
         # Tensor.to(device, dtype, non_blocking, copy, memory_format)
         # Tensor.to(other, non_blocking, copy)
         raise NotImplementedError(
             f"NF4Tensor.to({args[1:]}, {kwargs}) is not supported, passing to dispatch"
         )
+
+
```

Comment:
For cpu offloading, FSDP2 call `NF4Tensor.cpu()`. I have to implement `.cpu` in `__torch_function__` instead of `__torch_dispatch__`. In `__torch_dispatch__`, it will pass in default `dtype` (bf16,fp16,fp32) and we do not want to dequant to `dtype`